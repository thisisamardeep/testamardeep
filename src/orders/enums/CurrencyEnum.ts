export enum CurrencyEnum {
    EURUSD = "EURUSD",
    EURJPY = "EURJPY",
    AUDUSD = "AUDUSD",
    GBPUSD = "GBPUSD",
}
