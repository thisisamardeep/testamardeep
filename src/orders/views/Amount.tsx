import * as React from "react";

export const Amount: React.FC<AmountProps> = ({
                                                  value,
                                                  onAmountChanged,

                                              }) => {

    return (
        <React.Fragment>
            <p>
                <label>Amount:</label>
                <input
                    type="text"
                    value={value}
                    onChange={(event) => {
                        onAmountChanged(event);
                    }}
                />
            </p>
        </React.Fragment>
    )
};

export interface AmountProps {
    value: string;
    onAmountChanged: (e: React.ChangeEvent<HTMLInputElement>) => void,
}
