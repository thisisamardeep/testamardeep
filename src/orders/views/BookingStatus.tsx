import * as React from "react";

export const BookingStatus: React.FC<BookingStatus> = ({
                                                           status,
                                                       }) => {
    return (
        <React.Fragment>
            <h2>{status}</h2>
        </React.Fragment>
    )
};

export interface BookingStatus {
    status: string;
}
