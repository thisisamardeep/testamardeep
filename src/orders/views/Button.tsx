import * as React from "react";

export const Button: React.FC<ButtonProps> = ({
                                                  onClick,

                                              }) => {
    return (
        <React.Fragment>
            <p>
                <button onClick={(event) => {
                    onClick(event);
                }}>
                    Book
                </button>
            </p>
        </React.Fragment>
    )
};

export interface ButtonProps {
    onClick: (event: React.MouseEvent<HTMLButtonElement>) => void,
}
