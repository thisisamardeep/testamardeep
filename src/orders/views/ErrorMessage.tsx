import * as React from "react";

export const ErrorMessage: React.FC<ErrorProps> = ({
                                                       errorMessage,
                                                   }) => {


    return (
        <React.Fragment>
            <div>{errorMessage}</div>
        </React.Fragment>
    )

};

export interface ErrorProps {
    errorMessage: string;
}
