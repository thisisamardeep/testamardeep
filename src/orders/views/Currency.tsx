import * as React from "react";
import {CurrencyEnum} from "../enums/CurrencyEnum";

export const Currency: React.FC<CurrencyProps> = ({
                                                      currency,
                                                      onCurrencyPairChanged,

                                                  }) => {


    let options = [];
    for (let currencyEnumKey in CurrencyEnum) {
        if (CurrencyEnum.hasOwnProperty(currencyEnumKey)) {
            options.push(currencyEnumKey);
        }
    }

    return (
        <React.Fragment>
            <label>Currency:</label>
            <select value={currency} onChange={(e) => onCurrencyPairChanged(e)}>
                {options.map((value, index) => {
                    return <option value={value} key={index}>{value}</option>
                })}
            </select>
            <br/>
        </React.Fragment>
    )

};

export interface CurrencyProps {
    currency: CurrencyEnum;
    onCurrencyPairChanged: (e: React.ChangeEvent<HTMLSelectElement>) => void,
}
