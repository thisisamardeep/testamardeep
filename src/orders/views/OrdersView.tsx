import * as React from 'react';
import {Store} from '../infrastructure/Store';
import {Reducers} from '../store/Reducers';
import {SideEffects} from '../store/SideEffects';
import {CurrencyEnum} from '../enums/CurrencyEnum';
import {Amount} from "./Amount";
import {Currency} from "./Currency";
import {Button} from "./Button";
import {ACTION_AMOUNT_CHANGED, ACTION_BOOKING, ACTION_CURRENCYCHANGED} from "../store/actions";
import {ErrorMessage} from "./ErrorMessage";
import {BookingStatus} from "./BookingStatus";


interface IOrdersViewProps {
}

export interface IOrdersViewState {
    amount: string;
    currencyPair: CurrencyEnum;
    bookingRequest: {
        status: string;
        errorMessage: string;
    };
}

export default class OrdersView extends React.Component<IOrdersViewProps, IOrdersViewState> {
    private store: Store<IOrdersViewState>;

    constructor(props: IOrdersViewProps) {
        super(props);
        const initialState: IOrdersViewState = {
            amount: '10m',
            currencyPair: CurrencyEnum.EURJPY,
            bookingRequest: {
                status: null,
                errorMessage: null
            }
        };
        this.store = new Store(
            initialState,
            Reducers,
            SideEffects,
            (nextState: IOrdersViewState) => this.setState(nextState)
        );
        // set initial state
        this.state = this.store.currentState;
    }

    onAmountChanged = (e: React.ChangeEvent<HTMLInputElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_AMOUNT_CHANGED, e.target.value);
    };

    onCurrencyPairChanged = (e: React.ChangeEvent<HTMLSelectElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_CURRENCYCHANGED, e.target.value);
    };

    onBookRequested = (e: React.MouseEvent<HTMLInputElement>) => {
        e.preventDefault();
        this.store.dispatchAction(ACTION_BOOKING);
    };

    render() {
        const {amount, bookingRequest, currencyPair} = this.state;
        return (
            <div>
                <h1>OrdersList</h1>
                <Amount onAmountChanged={this.onAmountChanged} value={amount}/>
                <Currency currency={currencyPair}
                          onCurrencyPairChanged={this.onCurrencyPairChanged}/>
                <p>Order summary:</p>
                <p>Amount({amount})</p>
                <Button onClick={this.onBookRequested}/>
                <BookingStatus status={bookingRequest.status}/>
                <ErrorMessage errorMessage={bookingRequest.errorMessage}/>
            </div>
        );
    }
}
