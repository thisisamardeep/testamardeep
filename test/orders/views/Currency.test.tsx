import React from 'react';
import ReactDOM from 'react-dom';
import {act} from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import {Currency} from '../../../src/orders/views/Currency';
import {CurrencyEnum} from "../../../src/orders/enums/CurrencyEnum";

let container: any;

beforeEach(() => {
    // @ts-ignore
    window._virtualConsole.emit = jest.fn();
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});


describe('Currency', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<Currency currency={CurrencyEnum.EURUSD}
                              onCurrencyPairChanged={() => {
                              }}/>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('can render, submit and update', () => {
        act(() => {
            ReactDOM.render(<Currency currency={CurrencyEnum.EURUSD}
                                      onCurrencyPairChanged={() => {
                                      }}
            />, container);
        });
        const label = container.querySelector('label');
        expect(label.textContent).toBe('Currency:');
        const select = container.querySelector('select');
        expect(select.options.length).toBe(4);

    });
});
