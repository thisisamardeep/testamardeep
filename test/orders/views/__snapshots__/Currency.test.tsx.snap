// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency renders correctly 1`] = `
Array [
  <label>
    Currency:
  </label>,
  <select
    onChange={[Function]}
    value="EURUSD"
  >
    <option
      value="EURUSD"
    >
      EURUSD
    </option>
    <option
      value="EURJPY"
    >
      EURJPY
    </option>
    <option
      value="AUDUSD"
    >
      AUDUSD
    </option>
    <option
      value="GBPUSD"
    >
      GBPUSD
    </option>
  </select>,
  <br />,
]
`;
